expfmt/text_parse.go

```diff
@@ -492,7 +502,7 @@ func (p *TextParser) startTimestamp() stateFn {
 // readingHelp represents the state where the last byte read (now in
 // p.currentByte) is the first byte of the docstring after 'HELP'.
 func (p *TextParser) readingHelp() stateFn {
-	if p.currentMF.Help != nil {
+	if p.currentMF.GetHelp() != "" {
```

Comment:
Sorry for seeing this only now, but wouldn't this allow duplicate HELP strings as long as the first is empty? E.g. the following would be accepted with this PR:

```
# HELP foo
# HELP foo A help string.
# TYPE foo counter
foo 42
```

Maybe the check here now has to happen in line 201? I.e. if the help string is already non-nil there, create the parse error there and return nil. Otherwise, set it to empty and move on.

Reply:
I've been thinking about this case and I thought that perhaps we should only fail if the help string is none-empty.
So for your example 
```
# HELP foo
# HELP foo A help string.
# TYPE foo counter
foo 42
```

Would result in where the empty help string would be overwritten
```
# HELP foo A help string.
# TYPE foo counter
foo 42
```

I guess it's a matter of being strict about one and only one help entry.
If we decide that the parser should not override empty help values then we would need to move the check up to line 201 as you say.

Reply:
IIRC the parser is tolerant about _where_ in the file the line is (originally, that was meant to be valid exposition format). In turn, I think, that means we should be strict about duplicate lines. Otherwise, it would be confusing if you _want_ the help string to be empty. You add an explicit empty HELP line, but somewhere much further down, a rogue HELP line sets the help string to something else. It would appear like magic to the naive user.

Reply:
Ahhh I see what you mean. I did not think about this at all. I will implement your proposed changes.